### DISCUSSION TITLE
Can you import a file/macro by default without {% include ... %} definition?

### DISCUSSION THREAD

**Discussion by Andre601** [fralau/mkdocs-macros-plugin#190] [Q&A] [answered]
Would like to use this plugin to create a wiki-inspired setup to create info-boxes with data in it...

For that I would prefer to use jinja2 templating rather than having to touch python and setup entire code stuff...

I saw that I could import macros using the `{% include ... %}` jinja2 format, but I now was wondering if there was a way to define files and/or modules to include by default in a file to use, so that I don't have to do
```
{% include "macros/info_table.html" as info_table %}

{{ info_table.info(foo=bar) }}
```
but only
```
{{ info_table.info(foo=bar) }}
```

The reason for this is easier maintainability as I wouldn't need to update X files if I decide to change the file location or name...

I can do some python code, but I would prefer if the formatting/handling or variables would remain in a jinja2 template style...

**Comment by Andre601** [comment]
Would by any chance a snippet work here using pymdownx.snippet to include the `{% include ... %}` statement on pages by default?

**Comment by fralau** [comment]
Thanks for you input. This is something I have been thinking about... hypothetically. 🤔 

I covered this in https://github.com/fralau/mkdocs_macros_plugin/discussions/187#discussioncomment-7484138 :

> Why not having one auto-include template file; the plugin would insert that file (as a string) at the top of every markdown page that is being parsed?

> That file could be called `header.md` (it could be renamed). Then you would be able to define all the macros in it; or better (to separate your code and allow caching) put all the import directives into it.

> But not only; it could contain any markdown, HTML, or combination that one would need to have insert at the top of every markdown page. I already have a use case, where I would like to have a python macro to be executed at the top of every page. Currently, it is problematic, because I wish to insert it manually in each page.

There could be also a `footer.md` 🤔

**Reply by Andre601** [reply]
Okay.

In the mean time, what would be a good aproach to making this? I'm willing to make some python code, as long as it won't get too complicated in the end...
Also, markdown rendering would be preferred here for easier, wiki-like formatting.

**Comment by fralau** [comment, marked as answer]
Absolutely. For your needs, check the documentation on [post-production, for headers and footers](https://mkdocs-macros-plugin.readthedocs.io/en/latest/post_production/#application-to-headers-or-footers). That would go into your `module.py` file.

Instead of inserting a string, you could of course read a file.

**Reply by Andre601** [reply]
I don't quite get it...

I tried the example with adding a jinja2 directive but I get an error.

The python code I use (File is in `.hooks/macros.py` relative to mkdocs file):
```py
def on_pre_page_macros(env):
    footer = "\n\n{% import 'game.md' as game %}"
    env.markdown += footer
```

The error I receive:
```
INFO    -  [macros] - Found external Python module '.hooks/macros' in: D:\GitHub\chaosleaguewiki.github.io
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Scripts\mkdocs.exe\__main__.py", line 7, in <module>
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\click\core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\click\core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\click\core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\click\core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\click\core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\__main__.py", line 270, in serve_command
    serve.serve(**kwargs)
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\commands\serve.py", line 86, in serve
    builder(config)
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\commands\serve.py", line 67, in builder
    build(config, live_server=None if is_clean else server, dirty=is_dirty)
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\commands\build.py", line 277, in build
    config = config.plugins.on_config(config)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\plugins.py", line 527, in on_config
    return self.run_event('config', config)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\plugins.py", line 507, in run_event
    result = method(item, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs_macros\plugin.py", line 585, in on_config
    self._load_modules()
  File "C:\Users\******\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs_macros\plugin.py", line 458, in _load_modules
    self._load_module(module, local_module_name)
  File "C:\Users\*******\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs_macros\plugin.py", line 408, in _load_module
    raise NameError("No valid function found in module '%s'" %
NameError: No valid function found in module '.hooks/macros'
```

**Reply by fralau** [reply]
Thanks for the heads up. That's a misbehavior of the program, which requests the presence of the `define_env()` function in all cases.

See the workaround on #191 (you just need to add that function, even if empty).

**Comment by fralau** [comment]
Issue #191 was fixed.

**Comment by Andre601** [comment]
Another small thing I want to ask... Are there certain vars available through macros I define through a separate file that would allow access to page values?

I want to use the `page.title` as a fallback for an option, but this would require me to pass the `page` object to the macro itself, which is something I would like to avoid.

**Reply by Andre601** [reply]
Found an answer I assume: Using the `with context` in the import statement...

Tho, in this case is `page.title` return empty, even tho it isn't. Could it be because on when in the rendering the macros is called? Would a different var (Like `env`) exist for such a situation?

Will test and report back.

**Reply by Andre601** [reply]
I found out... Using `title=""` in the macro makes the text empty and the `d(page.title)` doesn't apply on empty text apparently...

Workaround I found is defining `page.title` int the `title=` itself.

**Comment by Andre601** [comment]
Not really related, but I want to ask here anyways: Is there a way to have the markdown parsed inside macros before it is applied?

I use the file aproach here, meaning the macros is a jinja2 macros I have as an md file and I would like that text that is part of an arg would be parsed before applied, so that I can have stuff like `` `something` `` rather than `<code>something</code>`

I assume this may not work because macros are done after the on_markdown call thing?

**Reply by fralau** [reply]
I am not sure I can fully follow you. I would need an example, since there are various possible cases.: you could use the `{% import ...%}` directive to import an external file as you did; or you could even call the Jinja2/macro interpreter from `main.py` if you wish.

**Reply by Andre601** [reply]
I just want to use markdown inside `{{ macros() }}` calls. Right now using something like this:
```
{{ game.info(
  inputs = "`!pull @user`",
  timer = "20 Seconds",
  rounds = "3",
  slots_guaranteed = "8",
  slots_raffle = "4",
  added = "v0.1 Alpha",
  
  img_url = "../../../assets/images/minigames/twitch/danger-zone.png"
) }}
```
results in this output, where the `!pull @user` keeps the backticks rather than being transformed into HTML `<code>` tags.
![grafik](https://github.com/fralau/mkdocs_macros_plugin/assets/11576465/9f589899-0aa3-44cb-92cd-827a8daa9bf4)

**Reply by fralau** [reply]
Impressive!

**Reply by fralau** [reply]
Hmm 🤔 , I would say yes. Of course, you can produce HTML from inside your macro written in Python, or render markdown into HTML . If you are happy with a simple markdown converter, just [use the canonic markdown package](https://python-markdown.github.io/). Here is a [tutorial](https://www.digitalocean.com/community/tutorials/how-to-use-python-markdown-to-convert-markdown-text-to-html). You can even complete it with extensions ("Mkdocs extensions" are actually [extensions of the `markdown` package](https://python-markdown.github.io/extensions/)).

If you wanted you could even make a recursive call to the full Mkdocs-Macros environment to convert/render your markdown + macros: write `env.render(my_code)` where `my_code`  is your string, and Bob's your uncle 🙂

**Reply by Andre601** [reply]
The issue is that this is a template file in jinja named `games.md`, so I would need to somehow send this content through the parser (again) to have the content parsed... I don't have it as a python code... which would probs look horrible anyways.

**Reply by Andre601** [reply]
Also, adding to this, just using the normal markdown parser and sending the content through `markdown.markdown(str)` gives an error because the macro is wrapped into `<p>` tags, which there seems to be an issue with as it is treated as "invalid syntax".

I also tried `env.render` like you suggested, but the output stays the same. Probs used it at the wrong timing? I used it in the `on_pre_page_macros` call, but I assume I should do it later (in the `define_env` perhaps?)... Or just on the macros content itself, which would mean I have to move everything over... And again, it ain't pretty in jinja, so it sure won't be in python either... And may even be more complicated tbh.

**Reply by fralau** [reply]
If it isn't pretty, it could be that you do not have the right approach: are you trying to make a macro do what a jinja2 template should do (laying out your markdown page with variable content)? Templates come first, macros come second (if templates are not sufficient).

If you want to do advanced stuff, it is important to familiarize yourself with the sequence of events in mkdocs, and what happens between them ([see the diagram in MkDocs](https://www.mkdocs.org/dev-guide/plugins/#events)). If it is too complicated, I suggest you leave it aside for the moment a go back to simpler stuff.

Suffice to say that using `env.render()` requires the right conditions. Macros are to be defined in the `define_env()` environment.

I suggest you try simple solutions first. Use templates first, macros second; and use standard markdown rendering.

**Reply by Andre601** [reply]
I already mentioned an issue with just using the markdown renderer, which is that the macro is treated as invalid due to being inserted in a `<p>` tag.

And since I don't have the macro in python can I not just put the options through the parser.
Tho a workaround may be making a simple `markdownify(text: str)` macro in python to parse stuff as inline markdown (if python markdown supports this, which I believe it does due to the attr_list extension).

**Reply by fralau** [reply]
That is strange:

```python
>>> from markdown import markdown
>>> markdown("Hello <p> **World**")
'<p>Hello <p> <strong>World</strong></p>'
```

**Reply by Andre601** [reply]
> That is strange:
> 
> ```python
> >>> from markdown import markdown
> >>> markdown("Hello <p> **World**")
> '<p>Hello <p> <strong>World</strong></p>'
> ```

Welcome to python markdown.

**Comment by Andre601** [comment]
Question related to the include statement...

In what position is it called? Like when I use `{% include ".icons/..." %}` in a macros of a template file I add using `include_dir` option, is the path relative to where the file is located? Or where the macros would be used?

Because I would like to use icons bundled in Material for MkDocs, but so far I get `TemplateNotFound` errors, most likely because the path isn't correct...
Is there a way to debug what path the macros currently has for the include statement? Does it support going back folders using `../`?

**Reply by Andre601** [reply]
Testing around the path seems to actually be relative to the location of the macro (Where it is stored).

**Reply by fralau** [reply]
See [indication about this in documentation.](https://mkdocs-macros-plugin.readthedocs.io/en/latest/advanced/#including-external-files-in-pages)
